libraries/lib-numeric-formats/formatters/BeatsNumericConverterFormatter.cpp

```diff
@@ -0,0 +1,234 @@
+      // For this reason we use (uts + 1) as the "range".
+      // On top of that, we want at least two digits to be shown. Audacity has a bug with the range of 11,
+      // where on digit will be calculated instead of 2.
+
+      auto& beatsField = mFields.emplace_back(
+         NumericField { false, 1, std::max(12, mUpperTimeSignature + 1), true });
+      beatsField.label = L" " + XO("beat").Translation();
+
+      if (mFracPart > 0)
+      {
+         beatsField.label += L" ";
+         // See the reasoning above about the range
+         auto& fracField = mFields.emplace_back(
+            NumericField { false, 1, std::max(12, mFracPart / mLowerTimeSignature + 1), true });
+      }
+
+      // Fill the aux mDigits structure
+      int pos = 0;
+      for (size_t i = 0; i < mFields.size(); i++)
+      {
+         mFields[i].pos = pos;
+
+         for (int j = 0; j < mFields[i].digits; j++)
+         {
+            mDigits.push_back(DigitInfo { int(i), j, pos });
+            pos++;
+         }
+
+         pos += mFields[i].label.length();
+      }
+
+      const auto beatLength = 60.0 / mTempo;
+      const auto barLength = mUpperTimeSignature * beatLength;
+      const auto fracLength = beatLength * mLowerTimeSignature / mFracPart;
+
+      mFieldLengths[0] = barLength;
+      mFieldLengths[1] = beatLength;
+      mFieldLengths[2] = fracLength;
+   }
+
+   void UpdateResultString(ConversionResult& result) const
+   {
+      for (size_t fieldIndex = 0; fieldIndex < mFields.size(); ++fieldIndex)
+      {
+         result.valueString +=
+            result.fieldValueStrings[fieldIndex] + mFields[fieldIndex].label;
+      }
+   }
+
+   ConversionResult ValueToString(double value, bool) const override
+   {
+      ConversionResult result;
+      result.fieldValueStrings.resize(mFields.size());
+
+      if (value < 0)
+      {
+         for (size_t fieldIndex = 0; fieldIndex < mFields.size (); ++fieldIndex)
+         {
+            const auto digitsCount = mFields[fieldIndex].digits;
+            auto& fieldValue = result.fieldValueStrings[fieldIndex];
+            for (int digitIndex = 0; digitIndex < digitsCount; ++digitIndex)
+               fieldValue += L"-";
+         }
+
+         UpdateResultString(result);
+
+         return result;
+      }
+
+      for (size_t fieldIndex = 0; fieldIndex < mFields.size(); ++fieldIndex)
+      {
+         const auto fieldLength = mFieldLengths[fieldIndex];
+         const auto fieldValue = static_cast<int>(std::floor(value / fieldLength));
+
+         result.fieldValueStrings[fieldIndex] =
+            wxString::Format(mFields[fieldIndex].formatStr, fieldValue + 1);
+
+         value = value - fieldValue * fieldLength;
+      }
+      
+      UpdateResultString(result);
+      return result;
+   }
+
+   std::optional<double> StringToValue(const wxString& valueString) const override
+   {      
+      if (
+         mFields.size() > 0 &&
+         valueString.Mid(mFields[0].pos, 1) == wxChar('-'))
```

Comment:
Did you mean `Mid` or `Left`?

Reply:
You copy-catted the old error in ParsedNumericConverterFormatter.cpp, if it is an error.

It happens not to cause an error in case all fields have at least two digits but it could be more general